Add basic resource estimation

gwwatkin opened the pull request:
Part of #227. Needs https://github.com/latticesurgery-com/web-ui/pull/117. 

Introduces some basic functionality to estimate how long and how many physical qubits are required to run the computation in question. Adds this data to the compilation text.

Including [OpenSurgery](https://github.com/alexandrupaler/opensurgery) as a git submodule to compute distillation box sizes.

gwwatkin committed:
Add a stub of a resource estimation class

gwwatkin commented on src/lsqecc/pipeline/lattice_surgery_compilation_pipeline.py:
```diff
@@ -71,4 +72,7 @@ def compile_str(
     )
 
+    # TODO| when compilation stages are supported, remove the 'Circuit|' from the text
+    compilation_text += "\nCircuit| " + estimate_resources(lsc.composer.getSlices()).render_ascii()
```
@Keelando , this line will conflict with #220, but it should be easy to fix. Ofc you can remove the comment and the `Circuit|` when merging 220. (probably want to keep the `\n` though)

gwwatkin replied:
I'm still debating whether we should keep the compilation text, drop it or turn it into a "Log" stage. Because it's very useful to debug in plain text.

gwwatkin committed:
Fix row/col typo

gwwatkin committed:
Integrate Qentiana

gwwatkin committed:
Add tests

gwwatkin committed:
Clone submodules and package opensurgery

gwwatkin commented on .github/workflows/test.yml:
```diff
@@ -31,6 +31,7 @@ jobs:
         
     - name: Build package
       run: |
+        git submodule update --init --recursive
```
@alexnguyenn  Looks like the CI is having problems fetching the submodules, thought this line would work ... do you have any ideas?

gwwatkin commented on .github/workflows/test.yml:
```diff
@@ -24,9 +24,10 @@ jobs:
       run: |
-        git submodule sync --recursive
+        git submodule init
+        git submodule update --recursive
```
Ok not sure what is going on here, this definitely work on my local machine

gwwatkin commented:
They really could try harder with error messages

gwwatkin commented:
Ok @alexnguyenn maybe you know what is going on ...  this makes pytest run on my machine on my machine, even with a newly cloned repo:
```
git submodule init
git submodule update --recursive
touch src/lsqecc/external/opensurgery/__init__.py
```
Though noticed that the checkout action does a recursive clone so the submodule part should not be necessary.

gwwatkin commented:
Ok @alexnguyenn ... made it eventually, please have a look at the github actions

alexnguyenn commented on .github/workflows/test.yml:
```diff
@@ -23,7 +23,12 @@ jobs:
-    
+      with:
+        submodules: recursive
+    - name: Make OpenSurgery into a package
+        run: |
+          touch src/lsqecc/external/opensurgery/__init__.py
```
```suggestion
    - name: Make OpenSurgery into a package
      run: |
        touch src/lsqecc/external/opensurgery/__init__.py
```
Alignment issue caused the failure I think

alexnguyenn commented:
I am confused here - what is this PR is doing and why are we adding opensurgery? Can you update the description with those infos?

I also just looked in the Github Actions log and seems like the test action didn't run at all.

gwwatkin committed:
Update .github/workflows/test.yml

Co-authored-by: Alex Nguyen <46719079+alexnguyenn@users.noreply.github.com>

gwwatkin commented:
@alexnguyenn, updated title and description. Had some problems with the github action ... squashed 10+ in which I was trying to fix it. The last one made no changes to the code, I think that's why the checks showed as not run, but you can see in the action log going a little back that they did go at one point before that (and many times they failed due to packaging problems :sweat_smile: ).

alexnguyenn commented on .github/workflows/test.yml:
```diff
@@ -62,6 +72,7 @@ jobs:
     
     - name: Test with pytest
       run: |
+        ls src/lsqecc/external/opensurgery
```
```suggestion
```
Don't think we need this.

gwwatkin committed:
Update .github/workflows/test.yml

Co-authored-by: Alex Nguyen <46719079+alexnguyenn@users.noreply.github.com>

gwwatkin committed:
Remove the adding of the __init__ file to open surgery
